testing/internal/e2e/tests/static/target_address_test.go

```diff
@@ -0,0 +1,340 @@
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	parts = strings.Fields(string(output.Stdout))
+	hostIp = parts[len(parts)-1]
+	require.Equal(t, c.TargetIp, hostIp, "SSH session did not return expected output")
+	t.Log("Successfully connected to target")
+}
+
+// TestCliTargetAddressToHostSource asserts that we can change a target using an
+// address to a target using host sources only if we remove the address first.
+// TestCliTargetAddressToHostSource and TestCliTargetHostSourceToAddress are separate,
+// independent tests because switching between host sources and address is not a
+// commutative operation (they use different codepaths).
+func TestCliTargetAddressToHostSource(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+	c, err := loadConfig()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+	newOrgId := boundary.CreateNewOrgCli(t, ctx)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary", e2e.WithArgs("scopes", "delete", "-id", newOrgId))
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+	newHostCatalogId := boundary.CreateNewHostCatalogCli(t, ctx, newProjectId)
+	newHostSetId := boundary.CreateNewHostSetCli(t, ctx, newHostCatalogId)
+	newHostId := boundary.CreateNewHostCli(t, ctx, newHostCatalogId, c.TargetIp)
+	boundary.AddHostToHostSetCli(t, ctx, newHostSetId, newHostId)
+	newTargetId := boundary.CreateNewAddressTargetCli(t, ctx, newProjectId, c.TargetPort, c.TargetIp)
+
+	// Connect to target and print host's IP address
+	output := e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"connect",
+			"-target-id", newTargetId,
+			"-exec", "/usr/bin/ssh", "--",
+			"-l", c.TargetSshUser,
+			"-i", c.TargetSshKeyPath,
+			"-o", "UserKnownHostsFile=/dev/null",
+			"-o", "StrictHostKeyChecking=no",
+			"-o", "IdentitiesOnly=yes", // forces the use of the provided key
+			"-p", "{{boundary.port}}", // this is provided by boundary
+			"{{boundary.ip}}",
+			"hostname", "-i",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	parts := strings.Fields(string(output.Stdout))
+	hostIp := parts[len(parts)-1]
+	require.Equal(t, c.TargetIp, hostIp, "SSH session did not return expected output")
+	t.Log("Successfully connected to target")
+
+	// Attempt add-host-sources. Should error because the target has an address.
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "add-host-sources",
+			"-id", newTargetId,
+			"-host-source", newHostSetId,
+			"-format", "json",
+		),
+	)
+	require.Error(t, output.Err, "Target add-host-sources did not return expected error")
+
+	var response boundary.CliError
+	require.NoError(t, json.Unmarshal(output.Stderr, &response))
+	require.Equal(t, http.StatusBadRequest, response.Status, "Expected error when adding host source to target with an address")
```

Comment:
I thought this was supposed to be `http.Conflict`? I don't hate `BadRequest` but just want to confirm.

Reply:
Yep, this is correct - @ddebko and I thought `BadRequest` made more sense here.

Reply:
OK!